libmscore/note.cpp

```diff
@@ -3122,6 +3122,36 @@ std::vector<Note*> Note::tiedNotes() const
       return notes;
       }
 
+//---------------------------------------------------------
+//   disconnectTiedNotes
+//---------------------------------------------------------
+
+void Note::disconnectTiedNotes()
+      {
+      if (tieBack() && tieBack()->startNote()) {
+            tieBack()->startNote()->remove(tieBack());
+            }
+      if (tieFor() && tieFor()->endNote()) {
+            tieFor()->endNote()->setTieBack(0);
+            }
+      }
+
+//---------------------------------------------------------
+//   connectTiedNotes
+//---------------------------------------------------------
+
+void Note::connectTiedNotes()
+      {
+      if (tieBack()) {
```

Comment:
Is it possible that tieBack() returns nullptr? If it is, whether it is a corrupted logic or usual case?
If it is a corrupted logic, it is better not to check nullptr since it allows finding the corrupted place faster thanks to the crash. It is better for development to have crashes on invalid program state, than checking nullptr and skipping such corrupted program state.

Reply:
Yes, note may have only a forward tie or no ties at all, and in those cases `tieBack()` will return `nullptr`. I see nothing wrong about it.

Reply:
Is calling connect/disconnect valid in the case of nullptr ties?

Reply:
Yes, the purpose of these member functions is just to ensure that all ties that should be connected to the given note will be properly connected (or disconnected) with this note and those notes that should be bound to it by the ties. The note is not required to have any of those ties, and if it doesn't have them then these functions will just do nothing.

In fact, these functions are based on the code that was placed in undo.cpp and chord.cpp to handle ties connection/disconnection when adding/deleting a single note. I moved it to separate functions to make note deletion process more unified (which was the purpose of the whole commit) and reduce code duplication, and made some changes to make it suitable for more general case that that of deletion and addition of the note to the chord. Although my "generalization" may be not so good, the discussed checks on existence of backward and forward ties were present in the previous version of this code too (see, e.g., [this](https://github.com/musescore/MuseScore/pull/3564/files#diff-d1625459d1ed3699c71c0ca942f53b09L624)), so I have not added anything when assuming that the note in the considered cases may have just one tie or no ties.